docs/tests/e2e-tests.md

```diff
@@ -0,0 +1,103 @@
+---
+title: End to End Tests
+---
+
+# End to End Tests
+
+[Cypress](https://www.cypress.io), a JavaScript based framework, is used for end
+to end (E2E) testing. Tests and associated utilities are within the `cypress`
+directory.
+
+```shell
+$ tree app/javascript/article-form -L 1
+cypress
+├── fixtures
+|   └── initialAdmin.json
+|   └── ...
+├── integration
+|   └── **/*.spec.js
+├── plugins
+└── support
+```
+
+In addition to Cypress, we use
+[cypress-testing-library](https://github.com/testing-library/cypress-testing-library).
+This provides custom Cypress commands and utilities to improve how E2E tests are
+written. dom-testing-library and preact-testing-library are already used for
+[front-end tests](frontend-tests.md), which both offer a similar API in the
+context of testing a Preact or Stimulus component. All these tools are part of
+the [Testing Library](https://testing-library.com) family.
+
+For the test web server, the
+[cypress-rails](https://github.com/testdouble/cypress-rails) gem is used to
+start a test web server that runs a rails test environemnt (`RAILS_ENV=test`).
+It also resets the databaase between test runs by starting a database
+transaction at the beginning of a test and perfors a rollback at the end of a
+test being run. The cypress-rails gem also provides a rake task that allows us
+to coordinate all this work.
+
+## Running E2E Tests Locally
+
+From the command line, run `yarn e2e` to start E2E testing.
+
+1. `bundle check` will run to ensure all the gems for the project are up to
+   date. If they are not, you will be prompted to run `bundle install`.
+
+```bash
+Doing a quick bundle check to make sure gems are all up to date.
+
+ * zonebie (0.6.1)
+Install missing gems with `bundle install`
+Unable to launch end to end tests. Ensure that all your gems are installed and up to date.
+```
+
+2. `yarn install` will run to ensure front-end packages are up to date.
+
+```bash
+[1/5] 🔍  Validating package.json...
+[2/5] 🔍  Resolving packages...
+success Already up-to-date.
+```
+
+3. You will be prompted to set up the end to end (E2E) test database. Type `y`
+   or `Y` to install the E2E test database. Typically this is a one time thing,
+   but it can be run as well if ever you corrupt your database.
+
+```bash
+Do you need to set up your end to end (E2E) testing database? Answer yes
+if this is your first time running E2E tests on your local machine or you
+need to recreate your E2E test database. (y/n)
+```
+
+4. If `y` or `Y` is pressed, the E2E test database will install.
+
+```bash
+Setting up the E2E database before running E2E tests...
+
+
+Dropped database 'PracticalDeveloper_test'
+Created database 'PracticalDeveloper_test'
+
+...
+```
+
+5. The Cypress test runner will open and you are now ready to run end to end
+   tests.
```

Comment:
Thought it might be useful to add a link here, in case people want to see a short walkthrough of the test runner controls/features

```suggestion
5. The [Cypress test runner](https://docs.cypress.io/guides/core-concepts/test-runner.html#Overview) will open and you are now ready to run end to end
   tests.
```